UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,698 @@
+        NodePtr = (FDT_NODE_HEADER *)((CONST CHAR8 *)Fdt + SubNode + Fdt32ToCpu (((FDT_HEADER *)Fdt)->OffsetDtStruct));
+        DEBUG ((DEBUG_INFO, "\n      SubNode(%08X)  %a", SubNode, NodePtr->Name));
+
+        if (AsciiStrnCmp (NodePtr->Name, GmaStr, AsciiStrLen (GmaStr)) == 0) {
+          DEBUG ((DEBUG_INFO, "  Found gma@ node \n"));
+          //
+          // Build Graphic info HOB .
+          //
+          GraphicsDev = BuildGuidHob (&gEfiGraphicsDeviceInfoHobGuid, sizeof (EFI_PEI_GRAPHICS_DEVICE_INFO_HOB));
+          ASSERT (GraphicsDev != NULL);
+          if (GraphicsDev == NULL) {
+            break;
+          }
+
+          SetMem (GraphicsDev, sizeof (EFI_PEI_GRAPHICS_DEVICE_INFO_HOB), 0xFF);
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "vendor-id", &TempLen);
+          ASSERT (TempLen > 0);
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            GmaID  = (UINT16)Fdt32ToCpu (*Data32);
+            DEBUG ((DEBUG_INFO, "\n   vendor-id"));
+            DEBUG ((DEBUG_INFO, "  %016lX\n", GmaID));
+
+            GraphicsDev->VendorId = GmaID;
+          }
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "device-id", &TempLen);
+          ASSERT (TempLen > 0);
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            GmaID  = (UINT16)Fdt32ToCpu (*Data32);
+            DEBUG ((DEBUG_INFO, "\n   device-id"));
+            DEBUG ((DEBUG_INFO, "  %016lX\n", GmaID));
+
+            GraphicsDev->DeviceId = GmaID;
+          }
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "revision-id", &TempLen);
+          ASSERT (TempLen > 0);
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            GmaID  = (UINT16)Fdt32ToCpu (*Data32);
+            DEBUG ((DEBUG_INFO, "\n   revision-id"));
+            DEBUG ((DEBUG_INFO, "  %016lX\n", GmaID));
+
+            GraphicsDev->RevisionId = (UINT8)GmaID;
+          }
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "subsystem-vendor-id", &TempLen);
+          ASSERT (TempLen > 0);
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            GmaID  = (UINT16)Fdt32ToCpu (*Data32);
+            DEBUG ((DEBUG_INFO, "\n   subsystem-vendor-id"));
+            DEBUG ((DEBUG_INFO, "  %016lX\n", GmaID));
+
+            GraphicsDev->SubsystemVendorId = GmaID;
+          }
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "subsystem-id", &TempLen);
+          ASSERT (TempLen > 0);
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            GmaID  = (UINT16)Fdt32ToCpu (*Data32);
+            DEBUG ((DEBUG_INFO, "\n   subsystem-id"));
+            DEBUG ((DEBUG_INFO, "  %016lX\n", GmaID));
+
+            GraphicsDev->SubsystemId = GmaID;
+          }
+        }
+
+        if (AsciiStrnCmp (NodePtr->Name, "espi", AsciiStrLen ("espi")) == 0) {

```

Comment:
please also support serial device under PCI scenario.

Reply:
currently espi is under PCI node.
    **pci-rb1@E8000000 {**
        compatible = "pci";
        bus-range = <0x008000df>;
        #address-cells = <0x00000003>;
        #size-cells = <0x00000002>;
        dma-ranges = <0xc1000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000001 0x00000000>;
        ranges = <0xc1000000 0x00000000 0xd0000000 0x00000000 0xd0000000 0x00000000 0x0e000000 0xc1800000 0x00000100 0x00000000 0x00000100 0x00000000 0x0000001f 0xe0000000 0x80800000 0x00000000 0x00002000 0x00000000 0x00002000 0x00000000 0x00006c00>;
        **espi@1F,0 {**
            #address-cells = [00 02];
            #size-cells = [00 01];
            compatible = "isa";
            serial@3F8 {
                compatible = "isa";
                reg-io-width = <0x00000001>;
                reg = <0x00000001 0x000003f8 0x00000008>;
                current-speed = <0x0001c200>;
            };
        };
    };

Reply:
with offline discussion with community, they prefer to use isa to support lpc and espi.
sorry for inconvenience but please help to revert back to isa (both example FDT table and the parser code)

Reply:
resolved